### PR TITLE
Task description

### PR DESCRIPTION
**Pull request by danylott**

### CODE REVIEW

File: app/iot/service.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+from .message import Message, MessageType
+
+
+def generate_id(length: int = 8):
+    return "".join(random.choices(string.ascii_uppercase, k=length))
+
+
+# Protocol is very similar to ABC, but uses duck typing
+# so devices should not inherit for it (if it walks like a duck, and quacks like a duck, it's a duck)
```
Comment (Abnormaltype) [inline app/iot/service.py:13]: good example 😁

File: app/iot/service.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+class Device(Protocol):
+    def connect(self) -> None:
+        ...  # Ellipsis - synonym to "pass", but sometimes has different meaning
```
Comment (Abnormaltype) [inline app/iot/service.py]: 😎

File: app/iot/service.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+def generate_id(length: int = 8):
+    return "".join(random.choices(string.ascii_uppercase, k=length))
+
+
+# Protocol is very similar to ABC, but uses duck typing
+# so devices should not inherit for it (if it walks like a duck, and quacks like a duck, it's a duck)
+class Device(Protocol):
+    def connect(self) -> None:
+        ...  # Ellipsis - similar to "pass", but sometimes has different meaning
```
Comment (Abnormaltype) [inline app/iot/service.py:16]: 😎

File: app/iot/service.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import random
+import string
+from typing import Protocol
+
+from .message import Message, MessageType
+
+
+def generate_id(length: int = 8):
```
Comment (Abnormaltype) [inline app/iot/service.py:8]:
```suggestion
def generate_id(length: int = 8) -> str:
```

**Review by Abnormaltype — APPROVED** [review]
LGTM

File: app/iot/devices.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import time
+
+from .message import MessageType
+
+
+# of course this code looks dumb, but imagine some real implementations of each method here
+class HueLightDevice:
+    def connect(self) -> None:
+        print("Connecting Hue Light.")
+        time.sleep(0.5)
```
Comment (Abnormaltype) [inline app/iot/devices.py:10]: make a constant

File: README.md [hunk 6]
```diff
@@ -1,5 +1,42 @@
...
+As you have noticed - it's fully synchronous and has blocking operations (sleep for example).
+The idea of this task is to make the program work faster (using asyncio) without breaking its logic.
+
+For easier asynchronous setup please follow these steps:
+1. Make your code use asyncio - connect/disconnect/send_message must be async functions (just the async/await interface). 
+   Is the code working faster? If the answer is no - then proceed.
+2. Try to `gather` registering devices - that should be easy, devices may be registered simultaneously.
+   And when you run the code - wow it works faster (at least should). 
+   That's because each device doesn't need to wait for the previous to register.
+3. After that `gather` running programs (in `service.py`) - here the code also may be easy.
```
Comment (alenatovstukha) [inline README.md:16]: 3. After that, `gather` running programs (in `service.py`) — here the code also may be easy. And if you run it — wow, it works so fast, `asyncio` is powerful! But look at the logic — we've broken it. We may start playing music on Speaker before it was turned on, same with flushing and cleaning Smart Toilet. So, despite it working fast, it works incorrectly.

File: README.md [hunk 7]
```diff
@@ -1,5 +1,42 @@
...
+   And if you run it - wow! it works so fast, asyncio is powerful. But look at the logic.
+   We've broken it: we may start playing music on Speaker before it was turned on, 
+   same with flushing and cleaning Smart Toilet. So, despite it working fast, it works incorrectly.
+4. And finally your task is to fix these logical problems. As you already have understood, some of these
```
Comment (alenatovstukha) [inline README.md:20]: 4. And finally, your task is to fix these logical problems. As you already have understood, some of these operations may be performed in "parallel", and some of them must still run in a sequence (flush -> clean). So, you may use these helper functions:

File: README.md [hunk 8]
```diff
@@ -1,5 +1,42 @@
...
+    
+    async def run_sequence(*functions: Awaitable[Any]) -> None:
+        for function in functions:
+            await function
+    
+    
+    async def run_parallel(*functions: Awaitable[Any]) -> None:
+        await asyncio.gather(*functions)
+    ```
+   And change the `wake_up_program`, and `sleep_program` blocks and their run - to a combination of
```
Comment (alenatovstukha) [inline README.md:35]: Also, change the `wake_up_program` and `sleep_program` blocks and their run — to the combination of the `run_sequence` and `run_parallel` sending messages to the service. Please do that separately, firstly for `wake_up`, then for `sleep`. So the logic won't be broken. After that, your code will work a bit longer (still faster than at the beginning), but at least it works correctly now.

**Review by alenatovstukha — APPROVED** [review]

**Comment by alenatovstukha** [comment]
проверь, пожалуйста, если я правильно повыделяля некоторые названия `кодом`
@danylott

**Pull request merged**